// @chainflip/rpc/dist/common.cjs
"use strict";Object.defineProperty(exports, "__esModule", {value: true});// src/common.ts
var _zod = require('zod');








































var _parserscjs = require('./parsers.cjs');

var rpcResult = {
  broker_request_swap_deposit_address: _parserscjs.brokerRequestSwapDepositAddress,
  broker_request_swap_parameter_encoding: _parserscjs.requestSwapParameterEncoding,
  cf_request_swap_parameter_encoding: _parserscjs.requestSwapParameterEncoding,
  cf_accounts: _parserscjs.cfAccounts,
  cf_account_info: _parserscjs.cfAccountInfo,
  cf_pool_depth: _parserscjs.cfPoolDepth,
  cf_boost_pools_depth: _parserscjs.cfBoostPoolsDepth,
  cf_environment: _parserscjs.cfEnvironment,
  cf_funding_environment: _parserscjs.cfFundingEnvironment,
  cf_ingress_egress_environment: _parserscjs.cfIngressEgressEnvironment,
  cf_pool_orders: _parserscjs.cfPoolOrders,
  cf_pool_price_v2: _parserscjs.cfPoolPriceV2,
  cf_pools_environment: _parserscjs.cfPoolsEnvironment,
  cf_supported_assets: _parserscjs.cfSupportedAssets,
  cf_swap_rate: _parserscjs.cfSwapRate,
  cf_swap_rate_v2: _parserscjs.cfSwapRateV2,
  cf_swap_rate_v3: _parserscjs.cfSwapRateV3,
  cf_swapping_environment: _parserscjs.cfSwappingEnvironment,
  chain_getBlockHash: _parserscjs.chainGetBlockHash,
  cf_boost_pool_details: _parserscjs.cfBoostPoolDetails,
  cf_boost_pool_pending_fees: _parserscjs.cfBoostPoolPendingFees,
  state_getMetadata: _parserscjs.stateGetMetadata,
  state_getRuntimeVersion: _parserscjs.stateGetRuntimeVersion,
  lp_total_balances: _parserscjs.lpTotalBalances,
  cf_failed_call_ethereum: _parserscjs.cfFailedCallEvm.nullable(),
  cf_failed_call_arbitrum: _parserscjs.cfFailedCallEvm.nullable(),
  cf_authority_emission_per_block: _parserscjs.numberOrHex,
  cf_epoch_duration: _zod.z.number(),
  cf_auction_state: _parserscjs.cfAuctionState,
  cf_monitoring_simulate_auction: _parserscjs.cfMonitoringSimulateAuction,
  cf_flip_supply: _parserscjs.cfFlipSuppy,
  cf_eth_state_chain_gateway_address: _parserscjs.ethereumAddress.nullable(),
  cf_eth_key_manager_address: _parserscjs.ethereumAddress.nullable(),
  cf_pool_orderbook: _parserscjs.cfPoolOrderbook,
  cf_get_trading_strategies: _parserscjs.cfGetTradingStrategies,
  cf_get_trading_strategy_limits: _parserscjs.cfGetTradingStrategyLimits,
  cf_available_pools: _parserscjs.cfAvailablePools,
  cf_safe_mode_statuses: _parserscjs.cfSafeModeStatuses,
  broker_encode_cf_parameters: _parserscjs.hexString,
  cf_encode_cf_parameters: _parserscjs.hexString,
  cf_oracle_prices: _parserscjs.cfOraclePrices,
  cf_lending_pools: _parserscjs.cfLendingPools,
  cf_lending_config: _parserscjs.cfLendingConfig,
  cf_loan_accounts: _parserscjs.cfLoanAccounts
};



exports.rpcResponse = _parserscjs.rpcResponse; exports.rpcResult = rpcResult;
