// @ston-fi/sdk/dist/chunk-LWB5HDJC.cjs
"use strict";Object.defineProperty(exports, "__esModule", {value: true}); var _class;

var _chunkLYVQCGUPcjs = require('./chunk-LYVQCGUP.cjs');


var _chunkYN73SU6Rcjs = require('./chunk-YN73SU6R.cjs');


var _chunkA5I2ZLUNcjs = require('./chunk-A5I2ZLUN.cjs');


var _chunkWF6EZ2CWcjs = require('./chunk-WF6EZ2CW.cjs');


var _chunkZJQ75LD3cjs = require('./chunk-ZJQ75LD3.cjs');

// src/contracts/dex/v2_2/router/CPIRouterV2_2.ts
var CPIRouterV2_2 = (_class = class extends _chunkA5I2ZLUNcjs.CPIRouterV2_1 {
  static __initStatic() {this.version = _chunkZJQ75LD3cjs.DEX_VERSION.v2_2}
  async getPool(provider, params) {
    const poolAddress = await this.getPoolAddressByJettonMinters(
      provider,
      params
    );
    return _chunkYN73SU6Rcjs.CPIPoolV2_2.create(poolAddress);
  }
  async getVault(provider, params) {
    const tokenMinter = provider.open(_chunkWF6EZ2CWcjs.JettonMinter.create(params.tokenMinter));
    const vaultAddress = await this.getVaultAddress(provider, {
      user: params.user,
      tokenWallet: await tokenMinter.getWalletAddress(this.address)
    });
    return _chunkLYVQCGUPcjs.VaultV2_2.create(vaultAddress);
  }
}, _class.__initStatic(), _class);



exports.CPIRouterV2_2 = CPIRouterV2_2;
