// @ston-fi/sdk/dist/chunk-UR7TJ24F.js
import {
  CPIRouterV2_1
} from "./chunk-22QM2JGC.js";
import {
  StableRouterV2_1
} from "./chunk-6UCAEONG.js";
import {
  StablePoolV2_1
} from "./chunk-OXC3GXIH.js";
import {
  WCPIRouterV2_1
} from "./chunk-5NJ6NT4D.js";
import {
  WCPIPoolV2_1
} from "./chunk-GMIWAM73.js";
import {
  WStableRouterV2_1
} from "./chunk-PCGZ3CGG.js";
import {
  WStablePoolV2_1
} from "./chunk-C7SF5WJK.js";
import {
  BaseRouterV2_1
} from "./chunk-QY6FACUK.js";
import {
  VaultV2_1
} from "./chunk-V2F4T45J.js";
import {
  CPIPoolV2_1
} from "./chunk-EMSCSI77.js";
import {
  BasePoolV2_1
} from "./chunk-AIIUO6JT.js";
import {
  LpAccountV2_1
} from "./chunk-3KKQOYNT.js";
import {
  PtonV2_1
} from "./chunk-5C3RWSYW.js";

// src/contracts/dex/v2_1/index.ts
var RouterV2_1 = class extends BaseRouterV2_1 {
  static CPI = CPIRouterV2_1;
  static Stable = StableRouterV2_1;
  static WCPI = WCPIRouterV2_1;
  static WStable = WStableRouterV2_1;
};
var PoolV2_1 = class extends BasePoolV2_1 {
  static CPI = CPIPoolV2_1;
  static Stable = StablePoolV2_1;
  static WCPI = WCPIPoolV2_1;
  static WStable = WStablePoolV2_1;
};
var DEX = {
  Router: RouterV2_1,
  Pool: PoolV2_1,
  LpAccount: LpAccountV2_1,
  Vault: VaultV2_1,
  pTON: PtonV2_1
};

export {
  RouterV2_1,
  PoolV2_1,
  DEX
};
