// @ston-fi/sdk/dist/chunk-45EJS233.cjs
"use strict";Object.defineProperty(exports, "__esModule", {value: true}); var _class; var _class2;

var _chunkA5I2ZLUNcjs = require('./chunk-A5I2ZLUN.cjs');


var _chunkCH6EIJH6cjs = require('./chunk-CH6EIJH6.cjs');


var _chunk7ENVDM26cjs = require('./chunk-7ENVDM26.cjs');


var _chunk42FVW2SWcjs = require('./chunk-42FVW2SW.cjs');


var _chunkGIDOXPICcjs = require('./chunk-GIDOXPIC.cjs');


var _chunk6JMZBVIJcjs = require('./chunk-6JMZBVIJ.cjs');


var _chunkGOIXG6S3cjs = require('./chunk-GOIXG6S3.cjs');


var _chunkJ6BWJHV3cjs = require('./chunk-J6BWJHV3.cjs');


var _chunkRPU6RIXMcjs = require('./chunk-RPU6RIXM.cjs');


var _chunkWQUCOAKWcjs = require('./chunk-WQUCOAKW.cjs');


var _chunkF6QEZUEHcjs = require('./chunk-F6QEZUEH.cjs');


var _chunkARDWDSLZcjs = require('./chunk-ARDWDSLZ.cjs');


var _chunkRRI66KHLcjs = require('./chunk-RRI66KHL.cjs');

// src/contracts/dex/v2_1/index.ts
var RouterV2_1 = (_class = class extends _chunkJ6BWJHV3cjs.BaseRouterV2_1 {
  static __initStatic() {this.CPI = _chunkA5I2ZLUNcjs.CPIRouterV2_1}
  static __initStatic2() {this.Stable = _chunkCH6EIJH6cjs.StableRouterV2_1}
  static __initStatic3() {this.WCPI = _chunk42FVW2SWcjs.WCPIRouterV2_1}
  static __initStatic4() {this.WStable = _chunk6JMZBVIJcjs.WStableRouterV2_1}
}, _class.__initStatic(), _class.__initStatic2(), _class.__initStatic3(), _class.__initStatic4(), _class);
var PoolV2_1 = (_class2 = class extends _chunkF6QEZUEHcjs.BasePoolV2_1 {
  static __initStatic5() {this.CPI = _chunkWQUCOAKWcjs.CPIPoolV2_1}
  static __initStatic6() {this.Stable = _chunk7ENVDM26cjs.StablePoolV2_1}
  static __initStatic7() {this.WCPI = _chunkGIDOXPICcjs.WCPIPoolV2_1}
  static __initStatic8() {this.WStable = _chunkGOIXG6S3cjs.WStablePoolV2_1}
}, _class2.__initStatic5(), _class2.__initStatic6(), _class2.__initStatic7(), _class2.__initStatic8(), _class2);
var DEX = {
  Router: RouterV2_1,
  Pool: PoolV2_1,
  LpAccount: _chunkARDWDSLZcjs.LpAccountV2_1,
  Vault: _chunkRPU6RIXMcjs.VaultV2_1,
  pTON: _chunkRRI66KHLcjs.PtonV2_1
};





exports.RouterV2_1 = RouterV2_1; exports.PoolV2_1 = PoolV2_1; exports.DEX = DEX;
