// @chainflip/utils/dist/addresses.cjs
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/addresses.ts
var addresses_exports = {};
__export(addresses_exports, {
  FLIP_CONTRACT_ADDRESSES: () => FLIP_CONTRACT_ADDRESSES,
  STATE_CHAIN_GATEWAY_ADDRESSES: () => STATE_CHAIN_GATEWAY_ADDRESSES
});
module.exports = __toCommonJS(addresses_exports);
var FLIP_CONTRACT_ADDRESSES = {
  backspin: "0x10C6E9530F1C1AF873a391030a1D9E8ed0630D26",
  sisyphos: "0xcD079EAB6B5443b545788Fd210C8800FEADd87fa",
  perseverance: "0xdC27c60956cB065D19F08bb69a707E37b36d8086",
  mainnet: "0x826180541412D574cf1336d22c0C0a287822678A"
};
var STATE_CHAIN_GATEWAY_ADDRESSES = {
  backspin: "0xeEBe00Ac0756308ac4AaBfD76c05c4F3088B8883",
  sisyphos: "0x1F7fE41C798cc7b1D34BdC8de2dDDA4a4bE744D9",
  perseverance: "0xA34a967197Ee90BB7fb28e928388a573c5CFd099",
  mainnet: "0x6995Ab7c4D7F4B03f467Cf4c8E920427d9621DBd"
};
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  FLIP_CONTRACT_ADDRESSES,
  STATE_CHAIN_GATEWAY_ADDRESSES
});
